File: frontend/lib/api/locations.ts
// API client for locations endpoints

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL;

// フロントエンド用インターフェース（既存のモックデータと互換性を保つ）
export interface AvailableFeatures {
  aiMeasurement: boolean;
  measurement: boolean;
  simulation: boolean;
  modelCreation: boolean;
  realTimeMonitoring?: boolean;
  riskAnalysis?: boolean;
  reportGeneration?: boolean;
  dataExport?: boolean;
  userManagement?: boolean;
  locationManagement?: boolean;
  [key: string]: boolean | undefined;
}

export interface Location {
  id: string;
  location_id?: string; // バックエンドのlocation_id
  name: string;
  region: string;
  prefecture: string;
  tunnelName: string;
  description: string;
  folderName: string;
  status: 'active' | 'monitoring' | 'completed' | 'planning';
  startDate: string;
  totalLength: number;
  progress: number;
  lastUpdated: string;
  measurementCount: number;
  alertLevel: 'normal' | 'warning' | 'danger';
  coordinates?: {
    lat: number;
    lng: number;
  };
  availableFeatures: AvailableFeatures;
}

// バックエンドレスポンス用インターフェース
export interface LocationAPIResponse {
  id: number;
  location_id: string | null;
  name: string;
  description: string | null;
  address: string | null;
  region: string | null;
  prefecture: string | null;
  tunnel_name: string | null;
  folder_name: string | null;
  status: 'active' | 'monitoring' | 'completed' | 'planning';
  start_date: string | null;
  total_length: number | null;
  progress: number; // Float field
  measurement_count: number;
  alert_level: 'normal' | 'warning' | 'danger';
  latitude: number | null;
  longitude: number | null;
  coordinates: {
    lat: number;
    lng: number;
  } | null;
  available_features: AvailableFeatures;
  lastUpdated: string;
  user_count: number;
  feature_count: number;
  enabled_feature_count: number;
  created_at: string;
  updated_at: string;
}

export class LocationsAPI {
  private baseUrl: string;

  constructor(baseUrl: string = API_BASE_URL) {
    this.baseUrl = baseUrl || 'http://localhost:8000/api/v1';
    console.log('LocationsAPI initialized with baseUrl:', this.baseUrl);
  }

  async getLocations(): Promise<Location[]> {
    console.log('Fetching locations from:', `${this.baseUrl}/locations`);
    try {
      const response = await fetch(`${this.baseUrl}/locations`);
      console.log('Response status:', response.status);
      if (!response.ok) {
        throw new Error(`Failed to fetch locations: ${response.status} ${response.statusText}`);
      }
      const data: LocationAPIResponse[] = await response.json();
      console.log('Raw API response:', data);
      return data.map(this.transformLocationResponse);
    } catch (error) {
      console.error('Error in getLocations:', error);
      throw error;
    }
  }

  async getLocation(id: string): Promise<Location> {
    const response = await fetch(`${this.baseUrl}/locations/${id}`);
    if (!response.ok) {
      throw new Error(`Failed to fetch location: ${response.statusText}`);
    }
    const data: LocationAPIResponse = await response.json();
    return this.transformLocationResponse(data);
  }

  // バックエンドのレスポンスをフロントエンド用の形式に変換
  private transformLocationResponse(apiResponse: LocationAPIResponse): Location {
    return {
      id: apiResponse.location_id || apiResponse.id.toString(), // フロントエンド用のIDはlocation_idを使用、なければid
      location_id: apiResponse.location_id,
      name: apiResponse.name,
      region: apiResponse.region || '',
      prefecture: apiResponse.prefecture || '',
      tunnelName: apiResponse.tunnel_name || '',
      description: apiResponse.description || '',
      folderName: apiResponse.folder_name || '',
      status: apiResponse.status,
      startDate: apiResponse.start_date || '',
      totalLength: apiResponse.total_length || 0,
      progress: apiResponse.progress || 0,
      lastUpdated: apiResponse.lastUpdated || apiResponse.updated_at,
      measurementCount: apiResponse.measurement_count || 0,
      alertLevel: apiResponse.alert_level,
      coordinates: apiResponse.coordinates,
      availableFeatures: apiResponse.available_features || {
        aiMeasurement: false,
        measurement: false,
        simulation: false,
        modelCreation: false,
      }
    };
  }
}

// Export a default instance
export const locationsAPI = new LocationsAPI();

// 地域ごとの拠点をグループ化
export function getLocationsByRegion(locations: Location[]): Record<string, Location[]> {
  return locations.reduce((acc, location) => {
    if (!acc[location.region]) {
      acc[location.region] = [];
    }
    acc[location.region].push(location);
    return acc;
  }, {} as Record<string, Location[]>);
}

// IDから拠点を取得
export function getLocationById(locations: Location[], id: string): Location | undefined {
  return locations.find(loc => loc.id === id);
}

// フォルダ名から拠点を取得
export function getLocationByFolderName(locations: Location[], folderName: string): Location | undefined {
  return locations.find(loc => loc.folderName === folderName);
}

// ステータスごとの拠点数を取得
export function getLocationStats(locations: Location[]) {
  const stats = {
    active: 0,
    monitoring: 0,
    completed: 0,
    planning: 0,
    total: locations.length
  };
  
  locations.forEach(loc => {
    stats[loc.status]++;
  });
  
  return stats;
}

// アラートレベルごとの拠点数を取得
export function getAlertStats(locations: Location[]) {
  const stats = {
    normal: 0,
    warning: 0,
    danger: 0
  };
  
  locations.forEach(loc => {
    stats[loc.alertLevel]++;
  });
  
  return stats;
}

File: microservices/ai_ameasure/app/api/endpoints/locations.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
from ...db.database import get_db
from ...crud import locations as crud_locations
from ...schemas.locations import LocationResponse, LocationCreate, LocationUpdate
from ...models.location import Location

router = APIRouter()


@router.get("/locations", response_model=List[LocationResponse])
def get_locations(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db)
):
    """拠点一覧取得"""
    locations = crud_locations.get_locations(db, skip=skip, limit=limit)
    
    # フロントエンド形式でレスポンスを構築
    response_data = []
    for location in locations:
        available_features = crud_locations.get_available_features(db, location.id)
        
        location_data = {
            "id": location.id,
            "location_id": location.location_id,
            "name": location.name,
            "description": location.description,
            "address": location.address,
            "region": location.region,
            "prefecture": location.prefecture,
            "tunnel_name": location.tunnel_name,
            "folder_name": location.folder_name,
            "status": location.status,
            "start_date": location.start_date,
            "total_length": location.total_length,
            "progress": location.progress,
            "measurement_count": location.measurement_count,
            "alert_level": location.alert_level,
            "latitude": location.latitude,
            "longitude": location.longitude,
            "coordinates": location.coordinates,
            "available_features": available_features,
            "lastUpdated": location.updated_at.isoformat(),
            "user_count": 0,  # TODO: 実装が必要
            "feature_count": len(available_features),
            "enabled_feature_count": sum(1 for enabled in available_features.values() if enabled),
            "created_at": location.created_at,
            "updated_at": location.updated_at,
        }
        response_data.append(location_data)
    
    return response_data


@router.get("/locations/{location_id}", response_model=LocationResponse)
def get_location(
    location_id: int,
    db: Session = Depends(get_db)
):
    """拠点詳細取得"""
    location = crud_locations.get_location(db, location_id=location_id)
    if location is None:
        raise HTTPException(status_code=404, detail="Location not found")
    
    available_features = crud_locations.get_available_features(db, location.id)
    
    location_data = {
        "id": location.id,
        "location_id": location.location_id,
        "name": location.name,
        "description": location.description,
        "address": location.address,
        "region": location.region,
        "prefecture": location.prefecture,
        "tunnel_name": location.tunnel_name,
        "folder_name": location.folder_name,
        "status": location.status,
        "start_date": location.start_date,
        "total_length": location.total_length,
        "progress": location.progress,
        "measurement_count": location.measurement_count,
        "alert_level": location.alert_level,
        "latitude": location.latitude,
        "longitude": location.longitude,
        "coordinates": location.coordinates,
        "available_features": available_features,
        "lastUpdated": location.updated_at.isoformat(),
        "user_count": 0,  # TODO: 実装が必要
        "feature_count": len(available_features),
        "enabled_feature_count": sum(1 for enabled in available_features.values() if enabled),
        "created_at": location.created_at,
        "updated_at": location.updated_at,
    }
    
    return location_data


@router.post("/locations", response_model=LocationResponse)
def create_location(
    location: LocationCreate,
    db: Session = Depends(get_db)
):
    """新規拠点作成"""
    db_location = crud_locations.create_location(db=db, location=location)
    available_features = crud_locations.get_available_features(db, db_location.id)
    
    location_data = {
        "id": db_location.id,
        "location_id": db_location.location_id,
        "name": db_location.name,
        "description": db_location.description,
        "address": db_location.address,
        "region": db_location.region,
        "prefecture": db_location.prefecture,
        "tunnel_name": db_location.tunnel_name,
        "folder_name": db_location.folder_name,
        "status": db_location.status,
        "start_date": db_location.start_date,
        "total_length": db_location.total_length,
        "progress": db_location.progress,
        "measurement_count": db_location.measurement_count,
        "alert_level": db_location.alert_level,
        "latitude": db_location.latitude,
        "longitude": db_location.longitude,
        "coordinates": db_location.coordinates,
        "available_features": available_features,
        "lastUpdated": db_location.updated_at.isoformat(),
        "user_count": 0,
        "feature_count": len(available_features),
        "enabled_feature_count": sum(1 for enabled in available_features.values() if enabled),
        "created_at": db_location.created_at,
        "updated_at": db_location.updated_at,
    }
    
    return location_data


@router.put("/locations/{location_id}", response_model=LocationResponse)
def update_location(
    location_id: int,
    location: LocationUpdate,
    db: Session = Depends(get_db)
):
    """拠点情報更新"""
    db_location = crud_locations.update_location(db=db, location_id=location_id, location=location)
    if db_location is None:
        raise HTTPException(status_code=404, detail="Location not found")
    
    available_features = crud_locations.get_available_features(db, db_location.id)
    
    location_data = {
        "id": db_location.id,
        "location_id": db_location.location_id,
        "name": db_location.name,
        "description": db_location.description,
        "address": db_location.address,
        "region": db_location.region,
        "prefecture": db_location.prefecture,
        "tunnel_name": db_location.tunnel_name,
        "folder_name": db_location.folder_name,
        "status": db_location.status,
        "start_date": db_location.start_date,
        "total_length": db_location.total_length,
        "progress": db_location.progress,
        "measurement_count": db_location.measurement_count,
        "alert_level": db_location.alert_level,
        "latitude": db_location.latitude,
        "longitude": db_location.longitude,
        "coordinates": db_location.coordinates,
        "available_features": available_features,
        "lastUpdated": db_location.updated_at.isoformat(),
        "user_count": 0,
        "feature_count": len(available_features),
        "enabled_feature_count": sum(1 for enabled in available_features.values() if enabled),
        "created_at": db_location.created_at,
        "updated_at": db_location.updated_at,
    }
    
    return location_data


@router.delete("/locations/{location_id}")
def delete_location(
    location_id: int,
    db: Session = Depends(get_db)
):
    """拠点削除"""
    success = crud_locations.delete_location(db=db, location_id=location_id)
    if not success:
        raise HTTPException(status_code=404, detail="Location not found")
    return {"message": "Location deleted successfully"}